lib/src/null_safety_analysis.dart

```diff
@@ -131,50 +139,60 @@ class NullSafetyAnalysis {
+                    'Could not analyze $fileUrl');
               }
               if (unitResult.isPart) continue;
               final languageVersionToken = unitResult.unit.languageVersionToken;
               if (languageVersionToken == null) continue;
-              if (Version(languageVersionToken.major,
-                      languageVersionToken.minor, 0) <
-                  _firstVersionWithNullSafety) {
-                return NullSafetyCompliance.notCompliant;
+              final languageVersion = LanguageVersion.fromLanguageVersionToken(
+                  languageVersionToken);
+              if (!languageVersion.supportsNullSafety) {
+                return NullSafetyAnalysisResult(
+                    NullSafetyCompliance.notCompliant,
+                    '$fileUrl is opting out of null safety');
               }
             }
           }
         }
-        return NullSafetyCompliance.compliant;
+        return NullSafetyAnalysisResult(NullSafetyCompliance.compliant, null);
       });
-      assert(packageInternallyGood != null);
-      if (packageInternallyGood == NullSafetyCompliance.analysisFailed) {
-        return NullSafetyCompliance.analysisFailed;
+      assert(packageInternalAnalysis != null);
+      if (packageInternalAnalysis.compliance ==
+          NullSafetyCompliance.analysisFailed) {
+        return packageInternalAnalysis;
       }
-      if (packageInternallyGood == NullSafetyCompliance.notCompliant) {
-        allPackagesGood = false;
+      if (packageInternalAnalysis.compliance ==
+          NullSafetyCompliance.notCompliant) {
+        firstBadPackage ??= packageInternalAnalysis;
       }
     }
-    if (allPackagesGood) return NullSafetyCompliance.compliant;
-    final rootLanguageVersion = _languageVersion(
-        await packageId.source.bind(_systemCache).describe(packageId));
-    if (rootLanguageVersion != null &&
-        rootLanguageVersion >= _firstVersionWithNullSafety) {
-      return NullSafetyCompliance.apiOnly;
-    }
-    return NullSafetyCompliance.notCompliant;
-  }
 
-  /// Returns the language version specified by the dart sdk
-  Version _languageVersion(Pubspec pubspec) {
-    final sdkConstraint = pubspec.sdkConstraints['dart'];
-    if (sdkConstraint is VersionRange) {
-      final rangeMin = sdkConstraint.min;
-      if (rangeMin == null) return null;
-      return Version(rangeMin.major, rangeMin.minor, 0);
+    if (firstBadPackage == null) {
+      return NullSafetyAnalysisResult(NullSafetyCompliance.compliant, null);
+    }
+    if (firstBadPackage.compliance == NullSafetyCompliance.analysisFailed) {
+      return firstBadPackage;
     }
-    return null;
+    return NullSafetyAnalysisResult(
+        NullSafetyCompliance.mixed, firstBadPackage.reason);
   }
 }
+
+class NullSafetyAnalysisResult {
+  final NullSafetyCompliance compliance;
+
+  /// `null` if compliance == [NullSafetyCompliance.compliant].
+  final String reason;
```

Comment:
Love it, we might even want to output this in the JSON output from `pub outdated`.